python/lsst/daf/butler/_butler.py

```diff
@@ -798,6 +799,83 @@ def get_dataset_type(self, name: str) -> DatasetType:
         """
         raise NotImplementedError()
 
+    @abstractmethod
+    def find_dataset(
+        self,
+        datasetType: DatasetType | str,
+        dataId: DataId | None = None,
+        *,
+        collections: CollectionArgType | None = None,
+        timespan: Timespan | None = None,
+        datastore_records: bool = False,
+        **kwargs: Any,
+    ) -> DatasetRef | None:
+        """Find a dataset given its `DatasetType` and data ID.
```

Comment:
Okay. This can in theory call `Butler._rewrite_data_id` now that it's moved up to Butler.

Reply:
I've done this and it works fine. It doesn't save us anything in `butler.get()` though because that does the day_obs/seq_num look up early so it can construct a proper DataId, and potentially expand it, before calling find_dataset.